### ISSUE TITLE
Sync buffer exhaustion problems

### ISSUE THREAD

**Issue by GoogleCodeExporter** [j-woz/exm-issues#666]
```
I've been looking at some scaling problems on the scicolsim benchmark and I've 
mostly narrowed it down to some sort of issue with servers getting blocked on 
server->server syncs.  My current guess is that some of the MPI_Sends in there 
are blocking because they can't be buffered. The messages are small enough to 
be eagerly sent, but there is a limit to the number that can be buffered at one 
time at the receiver.  I think if this is exceeded you can end up with 
deadlocks or long chains of servers waiting on each other.

I'm trying to confirm this with some profiling, and if confirmed there's a cray 
mpich seting to alter the default number of buffers I can use as a workaround.

Ultimately a nice solution would be to eagerly allocate a bunch of buffers to 
receive syncs and set up Irecvs.  If senders also used Isend and delayed 
cleaning up, I think this would avert the pathological behaviour.
```

Original issue reported on code.google.com by `tim.g.ar...@gmail.com` on 16 Apr 2014 at 4:24
- Merged into: #667
